paymentsheet/src/test/java/com/stripe/android/paymentsheet/IntentConfirmationDefinitionTest.kt

```diff
@@ -0,0 +1,407 @@
+import com.stripe.android.model.ConfirmPaymentIntentParams
+import com.stripe.android.model.ConfirmSetupIntentParams
+import com.stripe.android.model.PaymentIntentFixtures
+import com.stripe.android.model.PaymentMethod
+import com.stripe.android.model.PaymentMethodCreateParamsFixtures
+import com.stripe.android.model.PaymentMethodFixtures
+import com.stripe.android.model.PaymentMethodOptionsParams
+import com.stripe.android.model.SetupIntentFixtures
+import com.stripe.android.payments.paymentlauncher.InternalPaymentResult
+import com.stripe.android.payments.paymentlauncher.PaymentLauncher
+import com.stripe.android.payments.paymentlauncher.PaymentLauncherContract
+import com.stripe.android.paymentsheet.addresselement.AddressDetails
+import com.stripe.android.paymentsheet.addresselement.toConfirmPaymentIntentShipping
+import com.stripe.android.testing.FakePaymentLauncher
+import com.stripe.android.utils.FakeIntentConfirmationInterceptor
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+
+class IntentConfirmationDefinitionTest {
+    @Test
+    fun `'createLauncher' should call factory when creating launcher`() {
+        val launcher = FakePaymentLauncher()
+
+        val definition = createIntentConfirmationDefinition(
+            paymentLauncher = launcher,
+        )
+
+        val createdLauncher = definition.createLauncher(
+            activityResultCaller = mock {
+                on {
+                    registerForActivityResult<PaymentLauncherContract.Args, InternalPaymentResult>(any(), any())
+                } doReturn mock()
+            },
+            onResult = {}
+        )
+
+        assertThat(createdLauncher).isEqualTo(launcher)
+    }
+
+    @Test
+    fun `On 'action' with new payment method, should call 'IntentConfirmationInterceptor' with expected params`() =
+        runTest {
+            val intentConfirmationInterceptor = FakeIntentConfirmationInterceptor().apply {
+                enqueueCompleteStep()
+            }
+
+            val definition = createIntentConfirmationDefinition(
+                intentConfirmationInterceptor = intentConfirmationInterceptor,
+            )
+
+            definition.action(
+                confirmationOption = PaymentConfirmationOption.PaymentMethod.New(
+                    initializationMode = PaymentSheet.InitializationMode.PaymentIntent(
+                        clientSecret = "pi_123"
+                    ),
+                    createParams = PaymentMethodCreateParamsFixtures.DEFAULT_CARD,
+                    optionsParams = null,
+                    shippingDetails = AddressDetails(name = "John Doe"),
+                    shouldSave = true,
+                ),
+                intent = PaymentIntentFixtures.PI_REQUIRES_PAYMENT_METHOD,
+            )
+
+            assertThat(intentConfirmationInterceptor.calls.awaitItem()).isEqualTo(
```

Comment:
I'd suggest asserting on each individual field rather than on the entire object. I think that makes things more readable (e.g. you can assert that the client secret in the interceptor call is the same as the one in the confirmation option).

Also, when the test fails, it's easier to read the failure message

Reply:
You can also exclude fields that you don't need to test (though not sure if that's relevant here or not)